docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,294 @@
+
+1. The HPA should continue to support scaling based on percentage of CPU
+   request
+
+2. The HPA should support scaling on arbitrary metrics associated with
+   pods
+
+3. The HPA should support scaling on arbitrary metrics associated with
+   other Kubernetes objects in the same namespace as the HPA (and the
+   namespace itself)
+
+4. The HPA should make scaling on multiple metrics in a single HPA
+   possible and explicit (splitting metrics across multiple HPAs leads to
+   the possibility of fighting between HPAs)
+
+### Specification ###
+
+```go
+type HorizontalPodAutoscalerSpec struct {
+    // the target scalable object to autoscale
+    ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef"`
+
+    // the minimum number of replicas to which the autoscaler may scale
+    // +optional
+    MinReplicas *int32 `json:"minReplicas,omitempty"`
+    // the maximum number of replicas to which the autoscaler may scale
+    MaxReplicas int32 `json:"maxReplicas"`
+
+    // the metrics to use to calculate the desired replica count (the
+    // maximum replica count across all metrics will be used).  It is
+    // expected that any metrics used will decrease as the replica count
+    // increases, and will eventually increase if we decrease the replica
+    // count.
+    // +optional
+    Metrics []MetricSpec `json:"metrics,omitempty"`
+}
+
+// a type of metric source
+type MetricSourceType string
+var (
+    // a metric describing a kubernetes object
+    ObjectSourceType MetricSourceType = "object"
+    // a metric describing pods in the scale target
+    PodsSourceType MetricSourceType = "pods"
+    // a resource metric known to Kubernetes
+    ResourceSourceType MetricSourceType = "resource"
+)
+
+// a specification for how to scale based on a single metric
+// (only `type` and one other matching field should be set at once)
+type MetricSpec struct {
+    // the type of metric source (should match one of the fields below)
+    Type MetricSourceType `json:"type"`
+
+    // metric describing a single Kubernetes object
+    Object *ObjectMetricSource `json:"object,omitempty"`
+    // metric describing pods in the scale target
+    Pods *PodsMetricSource `json:"pods,omitemtpy"`
+    // resource metric describing pods in the scale target
+    // (guaranteed to be available and have the same names across clusters)
+    Resource *ResourceMetricSource `json:"resource,omitempty"`
+}
+
+// a metric describing a Kubernetes object
+type ObjectMetricSource struct {
+    // the described Kubernetes object
+    Target CrossVersionObjectReference `json:"target"`
+
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the target value of the metric (as a quantity)
+    TargetValue resource.Quantity `json:"targetValue"`
+}
+
+// metric describing pods in the scale target
+type PodsMetricSource struct {
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
```

Comment:
This is unclear to me - which pod is the metric from?  I guess you assume that the metric is an aggregate metric, but in that case, it's not clear if or how you ensure it has anything to do with the pods...  Is this just an arbitrary metric and target value?